From 95b98cda0e0649712110b1f2ba924e7175ffad70 Mon Sep 17 00:00:00 2001
From: Marcelo Tosatti <mtosatti@redhat.com>
Date: Thu, 22 May 2014 01:48:13 -0400
Subject: [virt] kvm/x86: disable master clock if TSC is reset during suspend

Message-id: <20140522014813.GA9603@amt.cnet>
Patchwork-id: 80100
O-Subject: [RHEL 7.1 PATCH] KVM: x86: disable master clock if TSC is reset during suspend
Bugzilla: 1098604
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

Updating system_time from the kernel clock once master clock
has been enabled can result in time backwards event, in case
kernel clock frequency is lower than TSC frequency.

Disable master clock in case it is necessary to update it
from the resume path.

Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 16a9602158861687c78b6de6dc6a79e6e8a9136f of
kvm/master branch)
BZ: 1098604
Brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=7486132

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index b72de32..f47ed4f 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -106,6 +106,8 @@ EXPORT_SYMBOL_GPL(kvm_max_guest_tsc_khz);
 static u32 tsc_tolerance_ppm = 250;
 module_param(tsc_tolerance_ppm, uint, S_IRUGO | S_IWUSR);
 
+static bool backwards_tsc_observed = false;
+
 #define KVM_NR_SHARED_MSRS 16
 
 struct kvm_shared_msrs_global {
@@ -1459,7 +1461,8 @@ static void pvclock_update_vm_gtod_copy(struct kvm *kvm)
      &ka->master_kernel_ns,
      &ka->master_cycle_now);
 
- ka->use_master_clock = host_tsc_clocksource & vcpus_matched;
+ ka->use_master_clock = host_tsc_clocksource && vcpus_matched
+    && !backwards_tsc_observed;
 
  if (ka->use_master_clock)
   atomic_set(&kvm_guest_has_master_clock, 1);
@@ -6751,6 +6754,7 @@ int kvm_arch_hardware_enable(void *garbage)
   */
  if (backwards_tsc) {
   u64 delta_cyc = max_tsc - local_tsc;
+  backwards_tsc_observed = true;
   list_for_each_entry(kvm, &vm_list, vm_list) {
    kvm_for_each_vcpu(i, vcpu, kvm) {
     vcpu->arch.tsc_offset_adjustment += delta_cyc;
-- 
1.7.1